File: backend/shared/src/domain/entities/ehr/medication.rs
//! EHR Medication Entity
//!
//! Corresponds to VistA File #52 (^PS) - Pharmacy/Prescriptions

use chrono::{DateTime, NaiveDate, Utc};
use serde::{Deserialize, Serialize};
use serde_json::Value as JsonValue;
use uuid::Uuid;

use crate::shared::AuditFields;

/// Medication status
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "VARCHAR")]
#[sqlx(rename_all = "lowercase")]
pub enum MedicationStatus {
    #[serde(rename = "active")]
    Active,
    #[serde(rename = "completed")]
    Completed,
    #[serde(rename = "discontinued")]
    Discontinued,
    #[serde(rename = "on_hold")]
    OnHold,
    #[serde(rename = "cancelled")]
    Cancelled,
}

impl Default for MedicationStatus {
    fn default() -> Self {
        MedicationStatus::Active
    }
}

/// Medication type
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "VARCHAR")]
#[sqlx(rename_all = "lowercase")]
pub enum MedicationType {
    #[serde(rename = "outpatient")]
    Outpatient,
    #[serde(rename = "inpatient")]
    Inpatient,
    #[serde(rename = "iv")]
    IV,
    #[serde(rename = "prn")]
    PRN,
}

impl Default for MedicationType {
    fn default() -> Self {
        MedicationType::Outpatient
    }
}

/// EHR Medication Entity
///
/// Represents a medication prescription/order.
/// Corresponds to VistA File #52 (^PS - Pharmacy).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EhrMedication {
    pub id: Uuid,

    /// Internal Entry Number (VistA IEN)
    pub ien: i64,

    /// Organization (multi-tenant)
    pub organization_id: Uuid,

    /// Patient ID
    pub patient_id: Uuid,

    /// Visit when ordered (optional)
    pub visit_id: Option<Uuid>,

    /// RxNorm code
    pub rxnorm_code: Option<String>,

    /// NDC code
    pub ndc_code: Option<String>,

    /// Drug name (generic)
    pub drug_name: String,

    /// Brand name (if applicable)
    pub brand_name: Option<String>,

    /// Dosage (e.g., "500mg")
    pub dosage: String,

    /// Dosage form (tablet, capsule, liquid, etc.)
    pub dosage_form: Option<String>,

    /// Route (oral, IV, topical, etc.)
    pub route: Option<String>,

    /// Frequency (e.g., "BID", "Q8H", "Daily")
    pub frequency: String,

    /// SIG (complete directions)
    pub sig: String,

    /// Quantity dispensed
    pub quantity: Option<i32>,

    /// Days supply
    pub days_supply: Option<i32>,

    /// Refills remaining
    pub refills_remaining: Option<i32>,

    /// Status
    pub status: MedicationStatus,

    /// Medication type
    pub medication_type: MedicationType,

    /// Start date
    pub start_date: NaiveDate,

    /// End date
    pub end_date: Option<NaiveDate>,

    /// Discontinue date
    pub discontinued_date: Option<NaiveDate>,

    /// Discontinue reason
    pub discontinued_reason: Option<String>,

    /// Prescribing provider
    pub prescriber_id: Option<Uuid>,

    /// Pharmacy (if filled)
    pub pharmacy: Option<String>,

    /// Instructions/notes
    pub instructions: Option<String>,

    /// MUMPS global data
    pub mumps_data: Option<JsonValue>,

    // Audit fields
    pub request_id: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub created_by: Option<Uuid>,
    pub updated_by: Option<Uuid>,
    pub system_id: Option<String>,
    pub version: i64,
}

impl EhrMedication {
    /// Create a new medication
    pub fn new(
        organization_id: Uuid,
        patient_id: Uuid,
        drug_name: String,
        dosage: String,
        frequency: String,
        sig: String,
    ) -> Self {
        let audit = AuditFields::new();
        Self {
            id: Uuid::new_v4(),
            ien: 0,
            organization_id,
            patient_id,
            visit_id: None,
            rxnorm_code: None,
            ndc_code: None,
            drug_name,
            brand_name: None,
            dosage,
            dosage_form: None,
            route: Some("oral".to_string()),
            frequency,
            sig,
            quantity: None,
            days_supply: None,
            refills_remaining: None,
            status: MedicationStatus::Active,
            medication_type: MedicationType::Outpatient,
            start_date: Utc::now().date_naive(),
            end_date: None,
            discontinued_date: None,
            discontinued_reason: None,
            prescriber_id: None,
            pharmacy: None,
            instructions: None,
            mumps_data: None,
            request_id: audit.request_id,
            created_at: audit.created_at,
            updated_at: audit.updated_at,
            created_by: audit.created_by,
            updated_by: audit.updated_by,
            system_id: audit.system_id,
            version: audit.version,
        }
    }

    /// Discontinue the medication
    pub fn discontinue(&mut self, reason: Option<String>) {
        self.status = MedicationStatus::Discontinued;
        self.discontinued_date = Some(Utc::now().date_naive());
        self.discontinued_reason = reason;
        self.updated_at = Utc::now();
    }

    /// Put medication on hold
    pub fn hold(&mut self) {
        self.status = MedicationStatus::OnHold;
        self.updated_at = Utc::now();
    }

    /// Resume a held medication
    pub fn resume(&mut self) {
        self.status = MedicationStatus::Active;
        self.updated_at = Utc::now();
    }

    /// Check if medication is currently active
    pub fn is_active(&self) -> bool {
        self.status == MedicationStatus::Active
    }

    /// Get display string (drug + dosage)
    pub fn display_name(&self) -> String {
        format!("{} {}", self.drug_name, self.dosage)
    }

    /// Get full prescription display
    pub fn full_display(&self) -> String {
        format!("{} {} - {}", self.drug_name, self.dosage, self.sig)
    }
}

File: cli/packages/libs/shared/src/schemas/ehr/medication.ts
/**
 * EHR Medication Schemas
 *
 * Runtime validation for medication records.
 * Corresponds to VistA File #52 (^PS)
 * Single source of truth - use z.infer<> for types.
 */

import { z } from "zod";
import { DateStringSchema, DateTimeSchema } from "../common";
import { createTypeGuard, createAssertion } from "../guards";

// ============================================================================
// Medication Enums
// ============================================================================

/**
 * Medication status schema
 */
export const EhrMedicationStatusSchema = z.enum([
  "active",
  "completed",
  "discontinued",
  "on_hold",
  "cancelled",
]);

export type EhrMedicationStatus = z.infer<typeof EhrMedicationStatusSchema>;

/**
 * Medication type schema
 */
export const EhrMedicationTypeSchema = z.enum(["outpatient", "inpatient", "iv", "prn"]);

export type EhrMedicationType = z.infer<typeof EhrMedicationTypeSchema>;

// ============================================================================
// Medication Schema
// ============================================================================

/**
 * Full EHR Medication schema
 */
export const EhrMedicationSchema = z.object({
  // Identity
  id: z.string().uuid(),
  ien: z.number().int().positive(),
  organizationId: z.string().uuid(),
  patientId: z.string().uuid(),
  visitId: z.string().uuid().optional(),

  // Drug identification
  rxnormCode: z.string().optional(),
  ndcCode: z.string().optional(),
  drugName: z.string().min(1, { message: "Drug name is required" }).max(300),
  brandName: z.string().max(300).optional(),

  // Dosage information
  dosage: z.string().min(1, { message: "Dosage is required" }).max(100),
  dosageForm: z.string().max(100).optional(),
  route: z.string().max(100).optional(),
  frequency: z.string().min(1, { message: "Frequency is required" }).max(100),
  sig: z.string().min(1, { message: "Sig (directions) is required" }).max(500),

  // Quantity and refills
  quantity: z.number().nonnegative().optional(),
  daysSupply: z.number().int().positive().optional(),
  refillsRemaining: z.number().int().nonnegative().optional(),

  // Status and type
  status: EhrMedicationStatusSchema,
  medicationType: EhrMedicationTypeSchema,

  // Dates
  startDate: DateStringSchema,
  endDate: DateStringSchema.optional(),
  discontinuedDate: DateStringSchema.optional(),
  discontinuedReason: z.string().max(500).optional(),

  // Prescriber and pharmacy
  prescriberId: z.string().uuid().optional(),
  pharmacy: z.string().max(200).optional(),
  instructions: z.string().max(1000).optional(),

  // Audit fields
  createdAt: DateTimeSchema.optional(),
  createdBy: z.string().optional(),
  updatedAt: DateTimeSchema.optional(),
  updatedBy: z.string().optional(),

  // MUMPS data
  mumpsData: z.record(z.string(), z.unknown()).optional(),
}).refine(
  // Business rule: If discontinued, must have discontinuedDate and reason
  data => {
    if (data.status === "discontinued") {
      return !!data.discontinuedDate && !!data.discontinuedReason;
    }
    return true;
  },
  {
    message: "Discontinued medications must have discontinuedDate and discontinuedReason",
    path: ["discontinuedDate"],
  }
).refine(
  // Business rule: If endDate exists, it must be after startDate
  data => {
    if (data.endDate) {
      return new Date(data.endDate) > new Date(data.startDate);
    }
    return true;
  },
  {
    message: "End date must be after start date",
    path: ["endDate"],
  }
);

export type EhrMedication = z.infer<typeof EhrMedicationSchema>;

// Type guards
export const isEhrMedication = createTypeGuard(EhrMedicationSchema);

// Assertions
export const assertEhrMedication = createAssertion(EhrMedicationSchema, 'EhrMedication');

// ============================================================================
// Create Medication Request Schema
// ============================================================================

/**
 * Create medication request validation
 */
export const CreateEhrMedicationRequestSchema = z.object({
  patientId: z.string().uuid(),
  visitId: z.string().uuid().optional(),
  rxnormCode: z.string().optional(),
  drugName: z.string().min(1, { message: "Drug name is required" }).max(300),
  dosage: z.string().min(1, { message: "Dosage is required" }).max(100),
  route: z.string().max(100).optional(),
  frequency: z.string().min(1, { message: "Frequency is required" }).max(100),
  sig: z.string().min(1, { message: "Sig (directions) is required" }).max(500),
  quantity: z.number().nonnegative().optional(),
  daysSupply: z.number().int().positive().optional(),
  refillsRemaining: z.number().int().nonnegative().optional(),
  startDate: DateStringSchema.optional(),
  instructions: z.string().max(1000).optional(),
});

export type CreateEhrMedicationRequest = z.infer<typeof CreateEhrMedicationRequestSchema>;

// Type guards
export const isCreateEhrMedicationRequest = createTypeGuard(CreateEhrMedicationRequestSchema);

// Assertions
export const assertCreateEhrMedicationRequest = createAssertion(
  CreateEhrMedicationRequestSchema,
  'CreateEhrMedicationRequest'
);

// ============================================================================
// Update Medication Request Schema
// ============================================================================

/**
 * Update medication request validation
 */
export const UpdateEhrMedicationRequestSchema = CreateEhrMedicationRequestSchema.partial().extend({
  id: z.string().uuid(),
  status: EhrMedicationStatusSchema.optional(),
  discontinuedReason: z.string().max(500).optional(),
});

export type UpdateEhrMedicationRequest = z.infer<typeof UpdateEhrMedicationRequestSchema>;

// Type guards
export const isUpdateEhrMedicationRequest = createTypeGuard(UpdateEhrMedicationRequestSchema);

// Assertions
export const assertUpdateEhrMedicationRequest = createAssertion(
  UpdateEhrMedicationRequestSchema,
  'UpdateEhrMedicationRequest'
);

// ============================================================================
// Medication Search Criteria Schema
// ============================================================================

/**
 * Medication search criteria validation
 */
export const EhrMedicationSearchCriteriaSchema = z.object({
  patientId: z.string().uuid().optional(),
  status: EhrMedicationStatusSchema.optional(),
  drugName: z.string().optional(),
  rxnormCode: z.string().optional(),
  prescriberId: z.string().uuid().optional(),
  activeOnly: z.boolean().optional(),
});

export type EhrMedicationSearchCriteria = z.infer<typeof EhrMedicationSearchCriteriaSchema>;

// Type guards
export const isEhrMedicationSearchCriteria = createTypeGuard(EhrMedicationSearchCriteriaSchema);

// ============================================================================
// Utility Functions
// ============================================================================

/**
 * Check if medication is active
 */
export function isMedicationActive(medication: EhrMedication): boolean {
  return medication.status === "active";
}

/**
 * Check if medication can be refilled
 */
export function canRefill(medication: EhrMedication): boolean {
  return (
    medication.status === "active" &&
    typeof medication.refillsRemaining === "number" &&
    medication.refillsRemaining > 0
  );
}

/**
 * Check if medication is discontinued
 */
export function isMedicationDiscontinued(medication: EhrMedication): boolean {
  return medication.status === "discontinued";
}

/**
 * Get remaining days supply
 */
export function getRemainingDaysSupply(medication: EhrMedication): number | null {
  if (!medication.daysSupply || !medication.startDate) {
    return null;
  }

  const start = new Date(medication.startDate);
  const now = new Date();
  const daysPassed = Math.floor((now.getTime() - start.getTime()) / (1000 * 60 * 60 * 24));
  const remaining = medication.daysSupply - daysPassed;

  return Math.max(0, remaining);
}
